src/hasSubObject.test.ts

```diff
@@ -126,47 +211,395 @@ describe("typing", () => {
+      // @ts-expect-error [ts2322] - nested sub-object has extra key `c`
+      hasSubObject({ a: { b: 1, c: 2 } })({ a: { b: 1 } });
+      // @ts-expect-error [ts2345] - nested sub-object has extra key `c`
+      pipe({ a: { b: 1 } }, hasSubObject({ a: { b: 1, c: 2 } }));
+
+      // @ts-expect-error [ts2379] - nested sub-object has extra key `c` (works with optional key)
+      hasSubObject({ a: { b: 1, c: 2 } })({ a: { b: 1 } } as {
+        a?: { b: number };
+      });
+      pipe(
+        { a: { b: 1 } } as { a?: { b: number } },
+        // @ts-expect-error [ts2345] - nested sub-object has extra key `c` (works with optional key)
+        hasSubObject({ a: { b: 1, c: 2 } }),
+      );
+    });
+
+    it("doesn't allow nested objects keys to have different types", () => {
+      // @ts-expect-error [ts2322] - nested sub-object has wrong value types
+      hasSubObject({ a: { b: 1, c: 2 } })({ a: { b: 4, c: "c" } });
+      // @ts-expect-error [ts2345] - nested sub-object has wrong value types
       pipe({ a: { b: 4, c: "c" } }, hasSubObject({ a: { b: 1, c: 2 } }));
+
+      // @ts-expect-error [ts2322] - deep-nested sub-object has wrong value types
+      hasSubObject({ a: { b: { c: "2" } } })({ a: { b: { c: 2 } } });
+      // @ts-expect-error [ts2345] - deep-nested sub-object has wrong value types
+      pipe({ a: { b: { c: 2 } } }, hasSubObject({ a: { b: { c: "2" } } }));
+
+      // @ts-expect-error [ts2322] - properly compares primitive and non-primitive types
+      hasSubObject({ a: { b: { c: 2 } } })({ a: { b: "a" } });
+      // @ts-expect-error [ts2345] - properly compares primitive and non-primitive types
+      pipe({ a: { b: "a" } }, hasSubObject({ a: { b: { c: 2 } } }));
+
+      // @ts-expect-error [ts2322] - properly compares primitive and non-primitive types
+      hasSubObject({ a: { b: "a" } })({ a: { b: { c: 2 } } });
+      // @ts-expect-error [ts2345] - properly compares primitive and non-primitive types
+      pipe({ a: { b: { c: 2 } } }, hasSubObject({ a: { b: "a" } }));
+    });
+
+    it("allows nested objects keys to have different types when types are overlapping", () => {
+      // ok - union type in nested super-object
+      hasSubObject({ a: { b: 1 } })({ a: { b: 1 as number | string } });
+      pipe({ a: { b: 1 as number | string } }, hasSubObject({ a: { b: 1 } }));
+
+      // ok - union type in nested sub-object
+      hasSubObject({ a: { b: 1 as number | string } })({ a: { b: 1 } });
+      pipe({ a: { b: 1 } }, hasSubObject({ a: { b: 1 as number | string } }));
+
+      // ok - union type in both nested super-object and sub-object
+      hasSubObject({
+        a: { b: 1 as number | string },
+      })({
+        a: { b: 1 as boolean | number },
+      });
+      pipe(
+        { a: { b: 1 as boolean | number } },
+        hasSubObject({ a: { b: 1 as number | string } }),
+      );
+    });
+
+    it("narrows with empty object", () => {
```

Comment:
**question:** can we rewrite these to be in a pipe? `filter(hasSubObject)` for example

Reply:
@cjquines Yes, we can rewrite it but in this case we won't be testing the negate type guard condition (`else`). Is it ok for data-last test cases?

Do you want something like this?

```tsx
const data = [] as Array<{ a?: string; b?: number }>;

pipe(data, filter(hasSubObject({ a: "a" })), (x) => {
  expectTypeOf(x).toMatchTypeOf<Array<{ a: string; b?: number }>>();
});
```

Reply:
`filter(isNot(hasSubObject))` would be the most likely usage in the field, i think?

Reply:
@cjquines It's not the same as the negate type guard case (`else` branch). `isNot` creates a new type guard. If someone needs to use data-last function with `else` condition then it won't be covered by tests though the case is quite possible. I think it's better to have tests for both. One tests the data-last function as it is. Another tests the data-last function with a real use case.

Reply:
fair enough. it'd be nice if we got the `isNot` case working, but this is fine for now